### ISSUE TITLE
SEGV during handshake when using very short deadline in client

### ISSUE THREAD

**Issue by dgquintas** [grpc/grpc#9039]
Using the new testcase introduced in #9019 for short deadlines, the following flake happens:

```
I1209 13:24:51.130053807  141237 short_deadlines.c:55]       test_invoke_short_deadline_request/chttp2/simple_ssl_with_oauth2_fullstack
D1209 13:24:51.162679284  141237 short_deadlines.c:118]      test with 4 ops, 15 ms deadline
E1209 13:24:51.188426144  141237 chttp2_server.c:140]        Handshaking failed: "No Error"
D1209 13:24:51.192381729  141493 security_handshaker.c:123]  Security handshake failed: {"created":"@1481318691.192274352","description":"Handshake read failed","file":"src/core/lib/security/transport/security_handshaker.c","file_line":232,"referenced_errors":[{"created":"@1481318691.192245621","description":"Socket closed","fd":10,"file":"src/core/lib/iomgr/tcp_posix.c","file_line":247,"target_address":"ipv6:[::1]:1026"}]}
ASAN:DEADLYSIGNAL
=================================================================
==141237==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x0000005bb1fa bp 0x7ffd7a7cfa40 sp 0x7ffd7a7cfa10 T0)
    #0 0x5bb1f9 in grpc_endpoint_shutdown /usr/local/google/home/dgq/grpc/forks/grpc/src/core/lib/iomgr/endpoint.c:58:7
    #1 0x71a1bb in security_handshaker_shutdown /usr/local/google/home/dgq/grpc/forks/grpc/src/core/lib/security/transport/security_handshaker.c:349:5
    #2 0x95ba22 in grpc_handshaker_shutdown /usr/local/google/home/dgq/grpc/forks/grpc/src/core/lib/channel/handshaker.c:59:3
    #3 0x95b941 in grpc_handshake_manager_shutdown /usr/local/google/home/dgq/grpc/forks/grpc/src/core/lib/channel/handshaker.c:148:5
    #4 0x75c4dc in chttp2_connector_shutdown /usr/local/google/home/dgq/grpc/forks/grpc/src/core/ext/transport/chttp2/client/chttp2_connector.c:103:5
    #5 0x748aa2 in grpc_connector_shutdown /usr/local/google/home/dgq/grpc/forks/grpc/src/core/ext/client_channel/connector.c:54:3
    #6 0x74c5f5 in disconnect /usr/local/google/home/dgq/grpc/forks/grpc/src/core/ext/client_channel/subchannel.c:270:3
    #7 0x74c3a3 in grpc_subchannel_unref /usr/local/google/home/dgq/grpc/forks/grpc/src/core/ext/client_channel/subchannel.c:285:5
    #8 0x9aeea9 in pf_destroy /usr/local/google/home/dgq/grpc/forks/grpc/src/core/ext/lb_policy/pick_first/pick_first.c:87:5
    #9 0x7490ae in grpc_lb_policy_weak_unref /usr/local/google/home/dgq/grpc/forks/grpc/src/core/ext/client_channel/lb_policy.c:98:5
    #10 0x9acdda in pf_connectivity_changed /usr/local/google/home/dgq/grpc/forks/grpc/src/core/ext/lb_policy/pick_first/pick_first.c:274:5
    #11 0x74f95f in on_external_state_watcher_done /usr/local/google/home/dgq/grpc/forks/grpc/src/core/ext/client_channel/subchannel.c:417:3
    #12 0x5bad61 in grpc_closure_run /usr/local/google/home/dgq/grpc/forks/grpc/src/core/lib/iomgr/closure.c:120:5
    #13 0x5c5cc1 in grpc_exec_ctx_flush /usr/local/google/home/dgq/grpc/forks/grpc/src/core/lib/iomgr/exec_ctx.c:71:9
    #14 0x5c651c in grpc_exec_ctx_finish /usr/local/google/home/dgq/grpc/forks/grpc/src/core/lib/iomgr/exec_ctx.c:109:3
    #15 0x6170f5 in grpc_channel_destroy /usr/local/google/home/dgq/grpc/forks/grpc/src/core/lib/surface/channel.c:353:3
    #16 0x579297 in shutdown_client /usr/local/google/home/dgq/grpc/forks/grpc/test/core/end2end/tests/short_deadlines.c:87:3
    #17 0x5787f6 in end_test /usr/local/google/home/dgq/grpc/forks/grpc/test/core/end2end/tests/short_deadlines.c:93:3
    #18 0x577200 in test_invoke_short_deadline_request /usr/local/google/home/dgq/grpc/forks/grpc/test/core/end2end/tests/short_deadlines.c:187:3
    #19 0x576f7e in short_deadlines /usr/local/google/home/dgq/grpc/forks/grpc/test/core/end2end/tests/short_deadlines.c:198:5
    #20 0x4dfae2 in grpc_end2end_tests /usr/local/google/home/dgq/grpc/forks/grpc/test/core/end2end/end2end_tests.c:412:7
    #21 0x4db373 in main /usr/local/google/home/dgq/grpc/forks/grpc/test/core/end2end/fixtures/h2_oauth2.c:235:5
    #22 0x7f1463700f44 in __libc_start_main /build/eglibc-oGUzwX/eglibc-2.19/csu/libc-start.c:287
    #23 0x419d65 in _start (/usr/local/google/home/dgq/grpc/forks/grpc/bins/asan/h2_oauth2_test+0x419d65)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV /usr/local/google/home/dgq/grpc/forks/grpc/src/core/lib/iomgr/endpoint.c:58:7 in grpc_endpoint_shutdown
==141237==ABORTING
FAILED: bins/asan/h2_oauth2_test short_deadlines GRPC_POLL_STRATEGY=epoll [ret=1, pid=141237]

```

See [here](https://gist.githubusercontent.com/dgquintas/8923404ef6b059c1062af9bc54f45d6f/raw/0a989440a6bd18f86d80dd1556db0c992a4206e2/gistfile1.txt) for the full log and run_tests command line.

**Comment by markdroth** [comment]
This looks very likely to be related to my new handshaker code.  It may be fixed by #8979, which I'm about to merge.

**Comment by dgquintas** [comment]
#8979 seems to fix it. Closing.
